### ISSUE TITLE
Slow queries when searching for nodes with more than 1 relationship of the same type

### ISSUE THREAD

**Issue by mrksph** [spring-projects/spring-data-neo4j#2212]
Hi, 

This is my domain entities:

```
abstract CustomNode {
    String type;
}
```
```
ParentOfClient  {
    @Id
    @GeneratedValue
    Long id;
    @TargetNode
    Client child;
}
```
```
ParentOfVirtual  {
    @Id
    @GeneratedValue
    Long id;
    @TargetNode
    Virtual child;
}
```
```
Client extends CustomNode {
    String name;
    @Relationship("PARENT_OF")
    List<ParentOfClient> clients;
    @Relationship("PARENT_OF")
    List<ParentOfVirtual> virtuals;
}
```

```
Virtual extends CustomNode {
    String vType;
    List<ParentOfClient> clients;
}
```

```
Parent {
    String codeV;
}
```

This is my data modelling:

A Parent node includes one Client 
`(p:Parent)-[INCLUDES]->(a:Client)`

A Client node can be parent of N Client nodes
`(a:Client)-[PARENT_OF]->(b:Client)`

A Client node can also be parent of N Virtual nodes (almost always, only 2 Virtual nodes)
`(a:Client)-[PARENT_OF]->(v:Virtual)`

A Virtual node can be parent of N Clients
`(v:Virtual)-[PARENT_OF]->(a:Client)`

When executing a parentRepository.findByCodeV("aCodeV") this is everything that SDN executes under the hood:

```
2021-04-05 18:49:12.057 DEBUG 18812 --- [nio-8080-exec-3] org.springframework.data.neo4j.cypher    : Executing:
MATCH (n:Parent) where n.codeV = $codeV return n.genType
2021-04-05 18:49:12.071 DEBUG 18812 --- [nio-8080-exec-3] org.springframework.data.neo4j.cypher    : Executing:
MATCH (n:`Parent`) WHERE n.codeV = $codeV WITH collect(id(n)) AS __sn__ RETURN __sn__
2021-04-05 18:49:12.077 DEBUG 18812 --- [nio-8080-exec-3] org.springframework.data.neo4j.cypher    : Executing:
MATCH (n:`Parent`) WHERE n.codeV = $codeV OPTIONAL MATCH (n:`Parent`)-[__sr__:`INCLUDES`]->(__srn__:`Client`:`Custom`) WITH collect(id(n)) AS __sn__, collect(id(__srn__)) AS __srn__, collect(id(__sr__)) AS __sr__ RETURN __sn__, __srn__, __sr__
2021-04-05 18:49:12.154 DEBUG 18812 --- [nio-8080-exec-3] org.springframework.data.neo4j.cypher    : Executing:
MATCH (n:`Client`:`Custom`) WHERE id(n) IN $__ids__ OPTIONAL MATCH (n)-[__sr__:`PARENT_OF`]->(__srn__:`Virtual`:`Custom`) WITH collect(id(n)) AS __sn__, collect(id(__srn__)) AS __srn__, collect(id(__sr__)) AS __sr__ RETURN __sn__, __srn__, __sr__
2021-04-05 18:49:12.180 DEBUG 18812 --- [nio-8080-exec-3] org.springframework.data.neo4j.cypher    : Executing:
MATCH (n:`Virtual`:`Custom`) WHERE id(n) IN $__ids__ OPTIONAL MATCH (n)-[__sr__:`PARENT_OF`]->(__srn__:`Client`:`Custom`) WITH collect(id(n)) AS __sn__, collect(id(__srn__)) AS __srn__, collect(id(__sr__)) AS __sr__ RETURN __sn__, __srn__, __sr__
2021-04-05 18:49:12.299 DEBUG 18812 --- [nio-8080-exec-3] org.springframework.data.neo4j.cypher    : Executing:
MATCH (n:`Client`:`Custom`) WHERE id(n) IN $__ids__ OPTIONAL MATCH (n)-[__sr__:`PARENT_OF`]->(__srn__:`Virtual`:`Custom`) WITH collect(id(n)) AS __sn__, collect(id(__srn__)) AS __srn__, collect(id(__sr__)) AS __sr__ RETURN __sn__, __srn__, __sr__
2021-04-05 18:49:12.373 DEBUG 18812 --- [nio-8080-exec-3] org.springframework.data.neo4j.cypher    : Executing:
MATCH (n:`Client`:`Custom`) WHERE id(n) IN $__ids__ OPTIONAL MATCH (n)-[__sr__:`PARENT_OF`]->(__srn__:`Client`:`Custom`) WITH collect(id(n)) AS __sn__, collect(id(__srn__)) AS __srn__, collect(id(__sr__)) AS __sr__ RETURN __sn__, __srn__, __sr__
2021-04-05 18:49:12.445 DEBUG 18812 --- [nio-8080-exec-3] org.springframework.data.neo4j.cypher    : Executing:
MATCH (n:`Client`:`Custom`) WHERE id(n) IN $__ids__ OPTIONAL MATCH (n)-[__sr__:`PARENT_OF`]->(__srn__:`Client`:`Custom`) WITH collect(id(n)) AS __sn__, collect(id(__srn__)) AS __srn__, collect(id(__sr__)) AS __sr__ RETURN __sn__, __srn__, __sr__
2021-04-05 18:49:12.452 DEBUG 18812 --- [nio-8080-exec-3] org.springframework.data.neo4j.cypher    : Executing:
MATCH (rootNodeIds) WHERE id(rootNodeIds) IN $rootNodeIds OPTIONAL MATCH ()-[relationshipIds]-() WHERE id(relationshipIds) IN $relationshipIds OPTIONAL MATCH (relatedNodeIds) WHERE id(relatedNodeIds) IN $relatedNodeIds WITH rootNodeIds AS n, collect(DISTINCT relationshipIds) AS __sr__, collect(DISTINCT relatedNodeIds) AS __srn__ RETURN n AS __sn__, __sr__, __srn__
```

I already reworked the class hierarchy to use @Node on abstract base classes and also to don't extend Concrete classes.
It got better but this simple query runs for a whole minute before returning the desired result.

I've already added more labels to further help SDN match the desired result but without success, it still goes for too long.

**Comment by mrksph** [comment]
Yes, with Spring Boot 2.4.1 the quieries SDN builds doesn't depends on OPTIONAL MATCH ing that does increase search time by too much.

**Comment by mrksph** [comment]
Hi, can we have any updates on this?

We are having problems because if we use SDN 6.0.4 or higher some basic queries start taking too much to complete.
If we use 6.0.4 or lower we don't get the Relationship's IDs when searching for Entities. If we try to use UUIDs instead, there are some cases where SDN is not saving them...

**Comment by meistermeier** [comment]
Yes, you are right: We moved in 6.0.4 from a path-based approach to a cascading match. On each level it would create the `OPTIONAL MATCH` for all possible relationships.
The previous method with the paths lead to a high amount of data that ramped up on the database side and lead to problems with responsiveness of the database. So some obvious very simple queries could more or less kill the database.
My assumption is that it is not the complexity of cyclic data in your case but the amount. Could you give us any figures of what we are talking about?

**Comment by mrksph** [comment]
Hi @meistermeier , yes you are right. The slow times are related to the amount of nodes of a given type.

We have 1 Million nodes in total. About 600k of the most common nodes.

**Comment by AlexVinogradov** [comment]
currently facing same issues with 5milion nodes in total. By any chance, is there a workaround to improve the response time?

**Comment by mrksph** [comment]
@AlexVinogradov 
Hi, in our case, we reviewed the Data Model we were using. 
To make it work what we did was design our Data Model different, in Neo4j we were trying to re-use the same Node Label in 3 distinct use cases. 
To make it work faster, we added 3 new labels added to the first one complementing it and making it possible to create a specific repository to that kind of model.

Example:

First, we had the "CustomHierarchy" node label. It was being used to represent 3 different things (3 different types of CustomHierarchy). 

Then we came up with this problem.

After redesigning our data model, we found that adding a specific label to each CustomHierarchy case was the best solution.

*Specific names are replaced for TypeA, TypeB, for privacy reasons...

(ch:CustomHierarchy:TypeA)
(ch:CustomHierarchy:TypeB)
(ch:CustomHierarchy:TypeC)

Then we would create a new Repository for that exact kind of CustomHierarchy instead of relying in the "base" type, which we did implement for practical reasons (some common methods for the 3 distinct types were moved here).

```
public interface CustomHierarchyRepository extends
        Neo4jRepository<BaseCustomHierarchyNode, Long>, CustomHierarchyCustomRepository {
. . .
}
```

```
public interface CustomHierarchyRepository extends
        Neo4jRepository<TypeA, Long>, CustomHierarchyCustomRepository {
. . .
}
```
```
public interface CustomHierarchyRepository extends Neo4jRepository<TypeB, Long>{
. . .
}
```
```
public interface CustomHierarchyRepository extends  Neo4jRepository<TypeC, Long>{
. . .
}
```

Hope this helps.
